README.md
# Software::LicenseMoreUtils #

This Perl modules provides some utilities on top of [Software::License](https://metacpan.org/pod/Software::License):

* The possibility to create
  [Software::License](https://metacpan.org/pod/Software::License)
  object using short names like `GPL-1`, `GPL_1`. This module accepts
  more variations on the short name than
  [new_from_short_name](https://metacpan.org/pod/Software::LicenseUtils#new_from_short_name)
  from
  [Software::LicenseUtils](https://metacpan.org/pod/Software::LicenseUtils).
* Support license summaries on Debian that redirect the user to the licenses available in
  `/usr/share/common-licenses`
* Provides a
  [LGPL-2](http://search.cpan.org/perldoc?Software%3A%3ALicense%3A%3ALGPL_2)
  object. LGPL-2 is deprecated but sometimes found in Debian packages.
* GPL summaries support the "or later" clause often used with GPL licenses

## Example ##

```perl
use Software::LicenseMoreUtils;
   
 my $lic = Software::LicenseMoreUtils->new_license_with_summary({
   short_name => 'Apache-2.0', # accepts also Apache-2 Apache_2_0
   holder => 'X. Ample'
});

# returns a license summary on Debian, returns license text elsewhere
my $text = $lic->summary_or_text;

# returns license full text
my $text = $lic->text;
```

## Compatibility ##

[new_from_short_name](http://search.cpan.org/perldoc?Software%3A%3ALicenseMoreUtils%3A%3ALicenseWithSummary#new_from_short_name)
has the same parameters as the
[new_from_short_name](https://metacpan.org/pod/Software::LicenseUtils#new_from_short_name)
provided by
[Software::License](https://metacpan.org/pod/Software::License). It
returns a new
[Software::LicenseMoreUtils::LicenseWithSummary](https://metacpan.org/pod/Software::LicenseMoreUtils::LicenseWithSummary)
object which has the same methods as
[Software::License](https://metacpan.org/pod/Software::License).

## What about RedHat and other Linux distributions ##

This module was written mainly to help Debian packaging. Adding
support for other distribution should be straightforward. PRs are
welcome.




Changes
{{$NEXT}}

1.009 2022-06-04

  * new_from_short_name(): holder parameter is now optional.
  * summary_or_text(): does not include copyright notice when
    holder parameter was not set when calling new_from_short_name()

1.008 2022-04-16

  * fix: handle SPDX short name like LGPL-3.0-only

1.007 2022-04-13

  * Add missing dependency on Software::License >= 0.103014

1.006 2022-04-10

  * new_from_short_name() can now create license object from
    SPDX v3 identifiers. I.e using short names like GPL-2.0-only
    or LGPL-2.1-or-later.
  * add spdx_expression to LGPL_2
  * require recent Test::More

1.005 2020-11-15

  * document that GPL licenses with "or any later version" are
    accepted. (e.g. GPL-2+ for GPL v 2 or any later version)
  
1.004 2018-11-02

  * fix CPAN tester tests: test full text instead of summary text
    on non debian system
  * fix full text extraction of LGPL_2 license

1.003 2018-11-01

  * add notice to summary (Debian #911403)
    Thanks to Felix Lechner for the patch
  * add GFDL_1_2 and GFDL_1_3 license summaries

1.002 2018-07-17

  * Fix LGPL2.1 summary

1.001 2018-07-06

  * add MPL short names (Debian #865472)

0.005 2018-06-29

  * deprecate new_license_with_summary constructor in favor
    of new_from_short_name
  * fix detection of Debian distro
  * add license_class method

0.004 2018-05-28

  * move LGPL_2 license in Software::LicenseMoreUtils

0.003 2018-05-26

  * add MPL_1_1 and MPL_2_0 summaries (Debian #865472)
  * add CC0_1_0 summary (Debian #887926)
  * fix examples in pod doc

0.002 2018-05-21

  * fix tests for non-Debian distributions. Tx Slaven (gh #1)
  * avoid undef warning when distro file does not exist

0.001 2018-05-20

  First version


lib/Software/LicenseMoreUtils/License.pm
package Software::LicenseMoreUtils::License;

use strict;
use warnings;
use 5.10.1;

use Path::Tiny;
use YAML::Tiny;
use List::Util qw/first/;
use Carp;
use Text::Template;

our $AUTOLOAD;

my %path_to_distro = (
    '/etc/debian_version' => 'debian',
);

my $distro_file = first { -e $_ } keys %path_to_distro;
my $distro = $path_to_distro{$distro_file} || 'unknown';


(my $module_file = __PACKAGE__.'.pm' ) =~ s!::!/!g;
my $yml_file = path($INC{$module_file})->parent->child("$distro-summaries.yml") ;

my $summaries = {} ;
if ($yml_file->is_file) {
    $summaries = YAML::Tiny->read($yml_file)->[0];
}

sub new {
    my ($class, $args) = @_;
    my $self = {
        license => $args->{license},
        or_later => $args->{or_later},
    };

    bless $self, $class;
}

sub summary {
    my $self = shift;

    my $later_text = $self->{or_later} ? ", or (at\nyour option) any later version" : '';

    (my $section_name = ref( $self->{license} )) =~ s/.*:://;
    my $summary = $summaries->{$section_name} // '';

    my $template = Text::Template->new(
        TYPE => 'STRING',
        DELIMITERS => [ qw({{ }}) ],
        SOURCE => $summary
    );

    return $template->fill_in(
        HASH => { or_later_clause => $later_text },
    );
}

sub debian_text {
    my $self = shift;
    carp "debian_text is deprecated, please use summary_or_text";
    return $self->summary || $self->fulltext;
}

sub summary_or_text {
    my $self = shift;
    return $self->summary || $self->fulltext;
}

sub AUTOLOAD {
    my $self = shift;
    my $lic = $self->{license};
    my ($sub) = ($AUTOLOAD =~ /(\w+)$/);
    return $lic->$sub(@_) if ($lic and $sub ne 'DESTROY');
}

1;



lib/Software/LicenseMoreUtils/LicenseWithSummary.pm
package Software::LicenseMoreUtils::LicenseWithSummary;

# ABSTRACT: Software::License with a summary

use strict;
use warnings;
use 5.10.1;

use Path::Tiny;
use YAML::Tiny;
use List::Util qw/first/;
use Carp;
use Text::Template;

our $AUTOLOAD;

# map location of distro file (like /etc/redhat_release) to distro.
# must match a <distro>_summaries.yml file in the directory containing
# this Perl module file
my %path_to_distro = (
    '/etc/debian_version' => 'debian',
);

my $distro_file = first { -e $_ } keys %path_to_distro;
my $distro = $path_to_distro{$distro_file // ''} || 'unknown';

(my $module_file = __PACKAGE__.'.pm' ) =~ s!::!/!g;
my $yml_file = path($INC{$module_file})->parent->child("$distro-summaries.yml") ;

my $summaries = {} ;
if ($yml_file->is_file) {
    $summaries = YAML::Tiny->read($yml_file)->[0];
}

sub new {
    my ($class, $args) = @_;
    my $self = {
        license => $args->{license},
        or_later => $args->{or_later},
        holder => $args->{holder},
    };

    return bless $self, $class;
}

sub distribution { return $distro }

sub summary {
    my $self = shift;

    my $later_text = $self->{or_later} ? ", or (at\nyour option) any later version" : '';

    (my $section_name = ref( $self->{license} )) =~ s/.*:://;
    my $summary = $summaries->{$section_name} // '';

    my $template = Text::Template->new(
        TYPE => 'STRING',
        DELIMITERS => [ qw({{ }}) ],
        SOURCE => $summary
    );

    return $template->fill_in(
        HASH => { or_later_clause => $later_text },
    );
}

sub license_class {
    my $self = shift;
    return ref($self->{license});
}

sub debian_text {
    my $self = shift;
    carp "debian_text is deprecated, please use summary_or_text";
    return $self->summary_or_text;
}

sub summary_or_text {
    my ($self) = @_;
    my $text;
    if (length $self->summary and $self->{holder}) {
        $text = join("\n",  grep { $_ } ($self->notice, $self->summary));
    }
    elsif (length $self->summary) {
        $text = $self->summary;
    }
    elsif ($self->{holder}) {
        $text =  $self->fulltext;
    }
    else {
        $text = $self->license;
    }
    return $text;
}

sub AUTOLOAD {
    my ($self, @args) = @_;
    my $lic = $self->{license};
    my ($sub) = ($AUTOLOAD =~ /(\w+)$/);
    if ($lic and $sub ne 'DESTROY') {
        return $lic->$sub(@args);
    }
    else {
        return;
    }
}

1;

__END__

=head1 SYNOPSIS

See L<Software::LicenseMoreUtils>. This class should be created with
  L<Software::LicenseMoreUtils/new_from_short_name>

=head1 DESCRIPTION

This module provides a wrapper around all C<Software::License::*> to add
a summary.

=head1 Methods

This class provides all the methods of the underlying
L<Software::License> object and the following methods.

=head2 summary

Returns the license summary, or an empty string.

=head2 summary_or_text

Returns the license summary or the text of the license. Like
L<Software::License/fulltext>, this method also returns the copyright
notice B<if> C<holder> parameter was set when calling
L<Software::LicenseMoreUtils/new_from_short_name>.

=head2 distribution

Returns the name of the Linux distribution found by this module. This
method is intended for tests or debugging.

=head2 license_class

Returns the Perl class name of the underlying L<Software::License> object.
E.g. C<Software::License::AGPL_3>

=cut




lib/Software/LicenseMoreUtils.pm
package Software::LicenseMoreUtils;

use strict;
use warnings;
use 5.10.1;

use Try::Tiny;
use Carp;
use Software::LicenseMoreUtils::LicenseWithSummary;
use Software::License 0.103014;

# ABSTRACT: More utilities and a summary for Software::License

use base qw/Software::LicenseUtils/;

# a short name with '+' at the end of the short name implies an
# "or later" clause.  i.e. GPL-1+ is "GPL-1 or any later version"
my %more_short_names = (
    'Apache-2'     => 'Software::License::Apache_2_0',
    'Artistic'     => 'Software::License::Artistic_1_0',
    'Artistic-1'   => 'Software::License::Artistic_1_0',
    'Artistic-2'   => 'Software::License::Artistic_2_0',
    'BSD-3-clause' => 'Software::License::BSD',
    'Expat'        => 'Software::License::MIT',
    'LGPL-2'       => 'Software::LicenseMoreUtils::LGPL_2',
    'LGPL_2'       => 'Software::LicenseMoreUtils::LGPL_2',
    'LGPL-3'       => 'Software::License::LGPL_3_0',
    'MPL-1.0'      => 'Software::License::Mozilla_1_0',
    'MPL-1.1'      => 'Software::License::Mozilla_1_1',
    'MPL-2.0'      => 'Software::License::Mozilla_2_0',

    # GPL SPDX identifiers have another convention for GPL version number
    'LGPL-2.0' => 'Software::LicenseMoreUtils::LGPL_2',

    'GPL-1.0'  => 'Software::License::GPL_1',
    'GPL-2.0'  => 'Software::License::GPL_2',
    'GPL-3.0'  => 'Software::License::GPL_3',
);

sub _create_license {
    my ( $class, $arg ) = @_;
    croak "no license short name specified"
          unless defined $arg->{short_name};

    my $lic_obj;
    try {
        $lic_obj = SUPER::new_from_short_name($arg);
    };

    return $lic_obj if $lic_obj;

    try {
        $lic_obj = SUPER::new_from_spdx_expression($arg);
    };

    return $lic_obj if $lic_obj;

    my $short = $arg->{short_name};
    $short =~ s/-(only|or-later)$//;
    $short =~ s/\+$//;

    my $subclass = $short;
    $subclass =~ s/[\-.]/_/g;

    my $info = $more_short_names{$short} || "Software::License::$subclass";
    my $lic_file = my $lic_class = $info;
    $lic_file =~ s!::!/!g;
    try {
        ## no critic (Modules::RequireBarewordIncludes)
        require "$lic_file.pm";
    } catch {
        Carp::croak "Unknow license with short name $short ($_)";
    } ;
    delete $arg->{short_name};
    # the holder default value fits well with BSD license text
    $lic_obj = $lic_class->new( { holder => 'the copyright holder', %$arg } );

    return $lic_obj;
}

sub new_license_with_summary {
    carp "new_license_with_summary is deprecated. Please use new_from_short_name";
    goto & new_from_short_name;
}

sub new_from_short_name {
    my ( $class, $arg ) = @_;
    croak "no license short name specified"
        unless defined $arg->{short_name};

    my $short = $arg->{short_name};

    my $info = $more_short_names{$short} || '';
    my $or_later = $short =~ /(\+|-or-later)$/ ? 1 : 0;
    my $lic = $class->_create_license($arg);

    my $xlic = Software::LicenseMoreUtils::LicenseWithSummary->new({
        license => $lic,
        or_later => $or_later,
        holder => $arg->{holder},
    });
    return $xlic;
}

1;

__END__

=head1 SYNOPSIS

 use Software::LicenseMoreUtils;

 my $lic = Software::LicenseMoreUtils->new_from_short_name({
    short_name => 'Apache-2.0', # or GPL-2+, Artistic-2 ...
    holder => 'X. Ample' # unlike Software::License, holder is optional
 });

 # On Debian, return a license summary, returns license text elsewhere
 # with or without copyright notice, depending if holder is set.
 my $text = $lic->summary_or_text;

 # returns license full text
 my $text = $lic->text;

=head1 DESCRIPTION

This module provides more utilities for L<Software::License>:

=over

=item *

Method L</new_from_short_name> returns a
L<Software::LicenseMoreUtils::LicenseWithSummary> object that provides all
functionalities of C<Software::License::*> objects and a summary on
some Linux distribution (see below).

=item *

L</new_from_short_name> accepts more short names than
L<Software::LicenseUtils>

=item *

L</new_from_short_name> accepts "or any later version" variant of GPL
licenses. When a short name like C<GPL-3+> is used, the license
summary contains "or (at your option) any later version" statement.

=item *

L<Software::License::LGPL-2> license is also provided. Even though
license C<LGPL-2.1> is preferred over C<LGPL-2>, some software in
Debian use C<LGPL-2>.

=back

=head1 License summary

In some distribution like Debian, all packages should come with the
full text of the licenses of the package software.

To avoid many duplication of long license text, the text of the most
common licenses are provided in C</usr/share/common-licenses>
directory. Then the license text of a package need only to provide a
summary of the license that refer to the location of the common
license.

All summaries are provided for Debian (so, for Ubuntu). Other
distributions are welcome to send pull request for their license
summaries.

=head1 Methods

=head2 new_from_short_name

 my $license_object = Software::LicenseMoreUtils->new_from_short_name({
      short_name => 'GPL-1', # mandatory
      holder => 'X. Ample' # optional
 }) ;

Unlike L<Software::License>, the C<Holder> parameter is optional. When
set, L<Software::LicenseMoreUtils::LicenseWithSummary/summary_or_text>
returns a copyright notice with the text of the summary of the
license.

Returns a new L<Software::LicenseMoreUtils::LicenseWithSummary> object
which is a L<Software::License> wrapped with a summary. This is a
drop-in replacement for the L<Software::License> object as all methods
are delegated to the underlying L<Software::License> object.

Known short license names are C<GPL-*>, C<LGPL-*> , and their "or
later version" variant C<GPL-*+>, C<LGPL-*+> C<Artistic> and
C<Artistic-*>. Unlike vanilla L<Software::License>, this module
accepts license name with "-" (e.g. C<GPL-2>) along with "_"
(e.g. "C<GPL_2>").

SPDX v3 identifiers can also be used as short names. I.e. short names
like C<GPL-2.0-only> or C<LGPL-2.1-or-later> are supported.

If the short name is not known, this method tries to create a license
object with C<Software::License> and the specified short name
(e.g. C<Software::License::MIT> with C<< short_name => 'MIT' >> or
C<Software::License::Apache_2_0> with C<< short_name => 'Apache-2.0' >>).

=cut

